gc/base/SparseAddressOrderedFixedSizeDataPool.cpp

```diff
@@ -0,0 +1,416 @@
+
+	/* Initialize _heapFreeList with initial values */
+	_heapFreeList = createNewSparseHeapFreeListNode(sparseHeapBase, _approxLargestFreeEntry);
+
+	if ((NULL == _freeListPool) || (NULL == _objectToSparseDataTable) || (NULL == _heapFreeList)) {
+		Trc_MM_SparseAddressOrderedFixedSizeDataPool_initialization_failure(sparseHeapBase, _freeListPool, _objectToSparseDataTable, _heapFreeList);
+		return false;
+	}
+
+	Trc_MM_SparseAddressOrderedFixedSizeDataPool_initialization_success(sparseHeapBase, _freeListPool, _objectToSparseDataTable, _heapFreeList);
+	return true;
+}
+
+void
+MM_SparseAddressOrderedFixedSizeDataPool::tearDown(MM_EnvironmentBase *env)
+{
+	if (NULL != _freeListPool) {
+		pool_kill(_freeListPool);
+		_freeListPool = NULL;
+	}
+
+	if (NULL != _heapFreeList) {
+		freeAllSparseHeapFreeListNodes();
+		_heapFreeList = NULL;
+	}
+
+	if (NULL != _objectToSparseDataTable) {
+		hashTableFree(_objectToSparseDataTable);
+		_objectToSparseDataTable = NULL;
+	}
+}
+
+/**
+ * Teardown and free the receiver by invoking the virtual #tearDown() function
+ */
+void
+MM_SparseAddressOrderedFixedSizeDataPool::kill(MM_EnvironmentBase *env)
+{
+	tearDown(env);
+	env->getForge()->free(this);
+}
+
+#if defined(OMR_GC_DOUBLE_MAPPING_FOR_SPARSE_HEAP_ALLOCATION)
+struct J9PortVmemIdentifier*
+MM_SparseAddressOrderedFixedSizeDataPool::findIdentifierForSparseDataPtr(void *dataPtr)
+{
+	MM_SparseDataTableEntry lookupEntry = MM_SparseDataTableEntry(dataPtr);
+	MM_SparseDataTableEntry *entry = (MM_SparseDataTableEntry *)hashTableFind(_objectToSparseDataTable, &lookupEntry);
+	if ((NULL == entry) || (entry->dataPtr != dataPtr)) {
+		Trc_MM_SparseAddressOrderedFixedSizeDataPool_findEntry_failure(dataPtr);
+		return NULL;
+	}
+
+	return entry->identifier;
+}
+
+void
+MM_SparseAddressOrderedFixedSizeDataPool::recordDoubleMapIdentifierForData(void *dataPtr, struct J9PortVmemIdentifier *identifier)
+{
+	MM_SparseDataTableEntry lookupEntry = MM_SparseDataTableEntry(dataPtr);
+	MM_SparseDataTableEntry *entry = (MM_SparseDataTableEntry *)hashTableFind(_objectToSparseDataTable, &lookupEntry);
+
+	if ((NULL == entry) || (entry->dataPtr != dataPtr)) {
+		Trc_MM_SparseAddressOrderedFixedSizeDataPool_findEntry_failure(dataPtr);
+	} else {
+		Trc_MM_SparseAddressOrderedFixedSizeDataPool_findEntry_success(dataPtr);
+		entry->identifier = identifier;
+	}
+}
+#endif /* OMR_GC_DOUBLE_MAPPING_FOR_SPARSE_HEAP_ALLOCATION */
+
+bool
+MM_SparseAddressOrderedFixedSizeDataPool::mapSparseDataPtrToHeapProxyObjectPtr(void *dataPtr, void* proxyObjPtr, uintptr_t size)
+{
+	MM_SparseDataTableEntry entry = MM_SparseDataTableEntry(dataPtr, proxyObjPtr, size);
+	void *result = hashTableAdd(_objectToSparseDataTable, &entry);
+
+	if (NULL == result) {
+		Trc_MM_SparseAddressOrderedFixedSizeDataPool_insertEntry_failure(dataPtr, (void *)size, proxyObjPtr);
+		return false;
+	}
+
+	Trc_MM_SparseAddressOrderedFixedSizeDataPool_insertEntry_success(dataPtr, (void *)size, proxyObjPtr);
+	return true;
+}
```

Comment:
Here and other similar places: Coding Standards recommend to avoid multi return functions. Would you please modify?